crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2615,7 +2615,18 @@ impl<'db> TypeInferenceBuilder<'db> {
                 if !dunder_getitem_method.is_unbound() {
                     return dunder_getitem_method
                         .call(self.db, &[slice_ty])
-                        .unwrap_with_diagnostic(self.db, value.as_ref().into(), self);
+                        .return_ty_result(self.db, value.as_ref().into(), self)
+                        .unwrap_or_else(|err| {
+                            self.add_diagnostic(
+                                (&**value).into(),
+                                "call-non-callable",
+                                format_args!(
+                                    "Method `__getitem__` is not callable on object of type '{}'.",
```

Comment:
It's possible that we'll want to incorporate more information here from `err`. The nice thing is we _can_ if we want.

Reply:
In fact I would like to :) I think the diagnostic should include the type of `__getitem__` that we found to not be callable. This could really help someone in debugging why their `__getitem__` isn't callable. I think that type is already available in `err`. I would suggest the phrasing `"Method __getitem__ of type {} is not callable on object of type {}."

Reply:
I guess this will require some decisions about how to handle the `UnionElement` and `UnionElements` cases; I think we could just use the full union type and omit the detail of which element(s) were not callable, in this case? Don't have strong feelings though, open to options.

Reply:
Sounds good, I'll play around with it. I want something flexible but not too verbose or onerous.

Reply:
Alright, I went with printing the full union type for now.